### PR TITLE
Application UI base

### LINKED ISSUE #7: Implement basic layout
Create general widgets, bind dummy commands. Grid with stretching. 
Main window: menu with buttons and entries, canvas with 'add' buttons;
Add text window: text redactor, font chooser, font parameters, color chooser, try and ok buttons, a canvas to draw text.

### PR DESCRIPTION
**Pull request by 1alexandra**
layout.py:
class Application constructing main window is added.
class TextConfigureApp constructing text configuring window is added.
Text redactor, text color chooser and text font choosed is added.
run.py:
running Application

closes #7

### CODE REVIEW

File: collage/layout.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
+import tkinter as tk
+from tkinter.colorchooser import askcolor
+from matplotlib import font_manager
+import numpy as np
+from datetime import datetime
```
Comment (Satbek) [inline collage/layout.py]: When we are packing application, we have only python3.7, python3.8 builtin libraries. Please add additional libraries to requirements.txt, and write tests for components which use them.

File: collage/layout.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+def get_system_fonts():
+    fonts = []
+    for x in font_manager.findSystemFonts():
+        x = x[::-1]
+        dot = x.find('.')
+        slash = x.find('\\')
+        x = x[slash-1:dot:-1]
+        fonts += [x]
+    return np.unique(fonts)
```
Comment (Satbek) [inline collage/layout.py]: move this function to some module and test it

File: collage/layout.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+
+class Application(tk.Frame):
+    def __init__(self, master=None):
+        super().__init__(master)
+        self.master = master
+        self.grid()
+        self.create_widgets()
+
+    def create_widgets(self):
```
Comment (Satbek) [inline collage/layout.py]: Add additional methods to initialise some widgets. This method is too big, and I think it should be separated.

File: collage/layout.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        
+    def open_text_window(self, event):
+        # Output: tk.canvas object
+        root = tk.Tk()
+        window = TextConfigureApp(master=root)
+        window.mainloop()
+        return window.get_return()
+        
+        
+class TextConfigureApp(tk.Frame):
```
Comment (Satbek) [inline collage/layout.py]: move this class to another file

File: collage/layout.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        
+        # Add and quit button
+        self.ok_button = tk.Button(buttons, text = "OK",
+                                   padx = 10, pady = 10, width = 10)
+        self.ok_button.grid()
+        self.ok_button.bind('<Button-1>', self.quit)
+        
+    def choose_color(self, event):
+        self.rgb, self.color_name = askcolor(parent = self, title = "Choose color:")
+        print(self.rgb, self.color_name)
```
Comment (Satbek) [inline collage/layout.py]: I think it would be better for logging
https://docs.python.org/3/library/logging.html

File: collage/run.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import tkinter as tk
+from layout import *
```
Comment (Satbek) [inline collage/run.py]: Is it necessary?

File: collage/layout.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        
+class TextConfigureApp(tk.Frame):
+    def __init__(self, master=None):
+        super().__init__(master)
+        self.master = master
+        self.rgb = (0, 0, 0)
+        self.grid()
+        self.create_widgets()
+
+    def create_widgets(self):
```
Comment (Satbek) [inline collage/layout.py]: Separate this method for:
* create_text_redactor
* create_font_chooser
* and so on

**Review by Satbek — CHANGES_REQUESTED** [review]

**Review by Satbek — APPROVED** [review]

File: collage/fonts.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+def get_system_fonts():
+    fonts = []
+    for x in font_manager.findSystemFonts():
+        x = x[::-1]
+        dot = x.find('.')
+        slash = x.find('\\')
+        x = x[slash-1:dot:-1]
+        fonts += [x]
+    return np.unique(fonts)
```
Comment (krchmark) [inline collage/fonts.py]: In my opinion, `set(fonts)` fits better here. `np.unique` is more for numerical types.

File: collage/fonts.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
+from matplotlib import font_manager
+import numpy as np
+
+
+def get_system_fonts():
+    fonts = []
+    for x in font_manager.findSystemFonts():
+        x = x[::-1]
+        dot = x.find('.')
+        slash = x.find('\\')
```
Comment (krchmark) [inline collage/fonts.py]: Are we supporting Linux?

File: collage/grid.py [hunk 10]
```diff
@@ -0,0 +1,7 @@
+def grid_frame(frame, frame_rows=[0], frame_cols=[0], origin_row=0, origin_column=0, origin_sticky='snwe'):
```
Comment (krchmark) [inline collage/grid.py]: Do we really need to receive a list here? Wouldn't a tuple of first and last rows be enough?

File: collage/mainwindow.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+
+class Application(tk.Frame):
+    def __init__(self, master=None):
+        super().__init__(master)
+        self.master = master
+        self.create_widgets()
+
+    def create_widgets(self):
+        grid_frame(self.master, [0], [0], None, None, None)
```
Comment (krchmark) [inline collage/mainwindow.py]: I think, you should use keyword arguments here to increase readability.

File: collage/mainwindow.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+        self.create_add_text_button(left_frame)
+        self.create_canvas_frame(right_frame)
+
+    def create_buttons(self, frame):
+        # TODO: bind command functions
+        # TODO: add icons
+        buttons_frame = tk.Frame(frame, bd=10)
+        grid_frame(buttons_frame, [0, 1, 2], [0, 1], 0, 0, 'news')
+        bw = 10
+        self.undo_button = tk.Button(buttons_frame, text='Undo', width=bw)
```
Comment (krchmark) [inline collage/mainwindow.py]: I think, we should define all these constant strings in some yaml/json.

File: collage/mainwindow.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+            button.grid(row=i // 2, column=i % 2, sticky='new')
+
+    def create_entries(self, frame):
+        # TODO: add validation functions
+        entries_frame = tk.Frame(frame, bd=10)
+        grid_frame(entries_frame, [0, 1, 2, 3], [0, 1], 1, 0, 'news')
+        px = 5
+
+        bw = 10
+        self.width_label = tk.Label(entries_frame, text="Width in pixels", padx=px)
```
Comment (krchmark) [inline collage/mainwindow.py]: Same about constant strings.

File: collage/mainwindow.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+        # TODO: bind command functions
+        # TODO: add icons
+        buttons_frame = tk.Frame(frame, bd=10)
+        grid_frame(buttons_frame, [0, 1, 2], [0, 1], 0, 0, 'news')
+        bw = 10
+        self.undo_button = tk.Button(buttons_frame, text='Undo', width=bw)
+        self.redo_button = tk.Button(buttons_frame, text='Redo', width=bw)
+        self.load_button = tk.Button(buttons_frame, text='Load model', width=bw)
+        self.save_button = tk.Button(buttons_frame, text='Save model', width=bw)
+        self.save2_button = tk.Button(buttons_frame, text='Save as...', width=bw)
```
Comment (krchmark) [inline collage/mainwindow.py]: Non-informative name of variable

Comment (krchmark) [reply collage/mainwindow.py]: Also, not sure if it is a good idea to define a separate variable for every button

File: collage/mainwindow.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+            entry.grid(row=i, column=1, sticky='w')
+
+        self.get_entry_button = tk.Button(entries_frame, text="Change parameters")
+        self.get_entry_button.grid(row=4, column=0, columnspan=2, sticky='new')
+        self.get_entry_button.bind('<Button-1>', self.change_canvas_parameters)
+
+    def create_add_text_button(self, frame):
+        text_button_frame = tk.Frame(frame, bd=10)
+        grid_frame(text_button_frame, [0], [0], 2, 0, 'news')
+        self.text_button = tk.Button(text_button_frame, text="Add text...", padx=5, pady=5)
```
Comment (krchmark) [inline collage/mainwindow.py]: It is a good practice to initialize all the fields in constructor (at least as None), just to understrand what are variables we are dealing with

**Review by krchmark — COMMENTED** [review]
Not finished yet, but there are some comments for now:
1) There are a lot of constants that defined right in the code. Not sure if it is okay.
2) Some parts are not so easy to understand. We should use more commentaries in the code.
3) Ideally interface should be totally customizable somewhere from the outside (including grid generation). But let's keep it for a better day.

File: collage/grid.py (hunk 10 above)
Comment (1alexandra) [reply collage/grid.py]: No. Here we define which rows/columns of frame will be stretchable. It is possible, that we don't want it to be neighboring.

File: collage/fonts.py (hunk 9 above)
Comment (Satbek) [reply collage/fonts.py]: yes

File: requirements.txt [hunk 16]
```diff
@@ -0,0 +1,2 @@
+matplotlib==3.2.1
+numpy==1.18.2
```
Comment (krchmark) [inline requirements.txt:2]: Not used anymore

**Pull request merged**